# Contents
## Importing Libraries
## User-defined functions
## Importing Data (orders_products_merged.pkl)
## If-Statements
## If-Statements with the loc() Function
### &emsp; High-range, mid-range, low-range products
## If-Statements with For-Loops
### &emsp; Busy day, regular day, least busy day
## Task 4.7 
### &emsp; Re-thinking 'busiest_day' column created above
### &emsp; Creating a 'busiest_period_of_day' column
### &emsp; Exporting as orders_products_busy.pkl


# Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# User-defined functions

In [2]:
def add_num(x,y):
    sum = x+y
    return sum

In [3]:
add_num(3, 4)

7

# Importing data

In [4]:
# Turning project folder into a string
path = r'C:\Users\davau\OneDrive - College of the Sequoias\Career Foundry\Data Immersion\Achievement 4 (Python)\Instacart Basket Analysis'

In [5]:
# Importing orders_products_merged.pkl
df_ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [6]:
df_ords_prods_merged.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  
0          0         Soda        77              7     9.0   both  
1          1         Soda        77              7     9.0   both  
2          1         Soda        77              7     9.0   both  
3          1         Soda        77              7     9.0   both  
4          1         Soda        77              7     9.0   both

In [7]:
# Creating a subset of first million rows of df_ords_prods_merged
df = df_ords_prods_merged[:1000000]

In [8]:
df.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  
0          0         Soda        77              7     9.0   both  
1          1         Soda        77              7     9.0   both  
2          1         Soda        77              7     9.0   both  
3          1         Soda        77              7     9.0   both  
4          1         Soda        77              7     9.0   both

In [9]:
df.tail()

order_id  user_id eval_set  order_number  orders_day_of_week  \
999995   2836489   164627    prior            14                   0   
999996   1843600   164632    prior             5                   1   
999997    733106   164632    prior             9                   2   
999998   1650124   164632    prior            17                   6   
999999   3378258   164632    prior            22                   6   

        order_hour_of_day  days_since_prior_order  product_id  \
999995                 15                     6.0       30489   
999996                 19                     9.0       30489   
999997                 22                    10.0       30489   
999998                 17                    13.0       30489   
999999                 10                    15.0       30489   

        add_to_cart_order  reordered     product_name  aisle_id  \
999995                  2          0  Original Hummus        67   
999996                  2          0  Original Hummus        67   
999997                  3          1  Original Hummus        67   
999998                  1          1  Original Hummus        67   
999999                  3          1  Original Hummus        67   

        department_id  prices _merge  
999995             20     7.5   both  
999996             20     7.5   both  
999997             20     7.5   both  
999998             20     7.5   both  
999999             20     7.5   both

# If-Statements

In [10]:
# Defining the function price_label
# This function categorizes items into price ranges: low, mid, and high.

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [11]:
# Running price_label on df.
# df['price_range'] stores the new labels in a new column in df called 'price_range'
# axis=1 runs the function on all rows; axis=0 would run it on all columns.

df['price_range'] = df.apply(price_label, axis=1)

C:\Users\davau\AppData\Local\Temp\ipykernel_2876\3789356307.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [12]:
df['price_range'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range, dtype: int64

In [13]:
df['prices'].max()

14.8

# If-Statements With the loc() Function

## Setting the conditions using loc()

In [14]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\davau\AppData\Local\Temp\ipykernel_2876\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [15]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [16]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

## Checking if the results are the same

In [17]:
df['price_range_loc'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

## Repeating the procedure on the entire df_ords_prods_merged

In [18]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [19]:
df_ords_prods_merged.loc[(df_ords_prods_merged['prices'] <= 15) & (df_ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [20]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [21]:
# Running value_counts on df_ords_prods_merged, out of curiosity.
df_ords_prods_merged['price_range_loc'].value_counts(dropna = False)

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: price_range_loc, dtype: int64

# If-Statements with For-Loops

In [22]:
# Simple for-loop check
for x in range(30, 45):
    print('My age is %d' % (x))

My age is 30
My age is 31
My age is 32
My age is 33
My age is 34
My age is 35
My age is 36
My age is 37
My age is 38
My age is 39
My age is 40
My age is 41
My age is 42
My age is 43
My age is 44


## Creating a "how busy is this day" column in df_ords_prods_merged

In [23]:
# Getting a frequency of 'orders_day_of_week' column
df_ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

Note that 0=Saturday, 1=Sunday, etc.

In [24]:
# Creating an empty list into which the values of our for-loop will be stored
result = []

In [25]:
# Filling 'result' with 'Busiest day', 'Least busy', or 'Regularly busy'
for value in df_ords_prods_merged["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [26]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [27]:
# Combining the 'result' list with df_ords_prods_merged
df_ords_prods_merged['busiest_day'] = result

In [28]:
df_ords_prods_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

# Task 4.7

## Step 1

The 'price_label' and 'busiest_day' columns have been created above.

## Step 2: re-thinking the 'busiest_day' column

Using the frequency table for 'orders_day_of_week' above, we see that the busiest days are Saturday and Sunday (days 0 and 1), and the slowest days are day Tuesday and Wednesday (days 3 and 4).

In [29]:
# Creating an empty list into which the values of our for-loop will be stored
result_2 = []

In [30]:
# Filling 'result_2' with 'Busiest', 'Least busy', or 'Regularly busy'
for x in df_ords_prods_merged["orders_day_of_week"]:
  if x == 0 or x == 1:
    result_2.append("Busiest")
  elif x == 3 or x == 4:
    result_2.append("Least busy")
  else:
    result_2.append("Regularly busy")

In [31]:
result_2

['Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Busiest',
 'Busiest',
 'Busiest',
 'Least busy',
 'Busiest',
 'Regularly busy',
 'Regularly busy',
 'Busiest',
 'Busiest',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Busiest',
 'Busiest',
 'Busiest',
 'Regularly busy',
 'Regularly busy',
 'Busiest',
 'Regularly busy',
 'Regularly busy',
 'Busiest',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Least busy',
 'Busiest',
 'Busiest',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest',
 'Busiest',
 'Regularly busy',
 'Busiest',
 'Busiest',
 'Least busy',
 'Regularly busy',
 'Busiest',
 'Busiest',
 'Busiest',
 'Busiest',
 'Busiest',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest',
 'Least busy',
 'Busiest',
 'Regularly busy',
 'Regularly busy',
 'Bus

In [32]:
# Combining the 'result_2' list with df_ords_prods_merged
df_ords_prods_merged['busiest_days'] = result_2

## Step 3: checking the newly-created 'busiest_days' column

In [33]:
# Frequency of 'busiest_days'
df_ords_prods_merged['busiest_days'].value_counts(dropna = False)

Regularly busy    12916111
Busiest           11864412
Least busy         7624336
Name: busiest_days, dtype: int64

The busiest days were days 0 and 1, with frequencies of 6204182 and 5660230, respectively.  The sum of those frequencies matches the frequency of rows labeled "Busiest," so the result of the 'busiest_days' column is correct.  The other two categories also add up.  So it seems that all went according to plan!

## Step 4: creating a 'busiest_period_of_day' column

In [34]:
# Getting a frequency of 'order_hour_of_day' column
df_ords_prods_merged['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

Since three categories (most orders, average orders, and fewest orders) are desired, I will split the 24 hours of the day into 8 hour blocks.  That is, 8 hours with the highest volume of sales will be labeled 'Most orders,' the 8 hours with the lowest volume will be labeled 'Fewest orders,' and the other 8 hours will be labeled 'Average orders.'

Note that the highest volume occurs from hours 9-16.  The lowest volume occurs from hours 23-6.  The middle volume occurs from hours 7-8 and 17-22.

In [35]:
# Creating an empty list into which the values of our for-loop will be stored
result_3 = []

In [36]:
# Filling 'result_3' with 'Most orders', 'Fewest orders', or 'Average orders'

for x in df_ords_prods_merged["order_hour_of_day"]:
  if x in range(9,17):                  #range will stop at 16, as desired
    result_3.append("Most orders")
  elif (x in range(0,7)) or x == 23:
    result_3.append("Fewest orders")
  else:
    result_3.append("Average orders")

In [37]:
result_3

['Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most ord

In [38]:
# Combining the 'result_3' list with df_ords_prods_merged
df_ords_prods_merged['busiest_period_of_day'] = result_3

## Step 5: printing the frequency of 'busiest_period_of_day'

In [39]:
# Frequency of 'busiest_period_of_day'
df_ords_prods_merged['busiest_period_of_day'].value_counts(dropna = False)

Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: busiest_period_of_day, dtype: int64

Again, the numbers add up here.

## Step 6: ensuring notebook is clean, structured, and commented

## Step 7: export the dataframe as a pickle file

In [40]:
# Exporting df_ords_prods_merged as orders_products_busy.pkl
df_ords_prods_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_busy.pkl'))